vllm/v1/serial_utils.py

```diff
@@ -1,61 +1,112 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import pickle
+from collections.abc import Sequence
+from inspect import isclass
 from types import FunctionType
-from typing import Any, Optional
+from typing import Any, Optional, Union
 
 import cloudpickle
+import numpy as np
 import torch
+import zmq
 from msgspec import msgpack
 
-CUSTOM_TYPE_TENSOR = 1
-CUSTOM_TYPE_PICKLE = 2
-CUSTOM_TYPE_CLOUDPICKLE = 3
+CUSTOM_TYPE_PICKLE = 1
+CUSTOM_TYPE_CLOUDPICKLE = 2
+
+bytestr = Union[bytes, bytearray, memoryview, zmq.Frame]
 
 
 class MsgpackEncoder:
-    """Encoder with custom torch tensor serialization."""
+    """Encoder with custom torch tensor and numpy array serialization."""
 
     def __init__(self):
-        self.encoder = msgpack.Encoder(enc_hook=custom_enc_hook)
-
-    def encode(self, obj: Any) -> bytes:
-        return self.encoder.encode(obj)
-
-    def encode_into(self, obj: Any, buf: bytearray) -> None:
-        self.encoder.encode_into(obj, buf)
+        self.encoder = msgpack.Encoder(enc_hook=self.enc_hook)
+        self.aux_buffers: Optional[list[bytestr]] = None
+
+    def encode(self, obj: Any) -> Sequence[bytestr]:
+        try:
+            self.aux_buffers = bufs = [b'']
+            bufs[0] = self.encoder.encode(obj)
+            return bufs
```

Comment:
```suggestion
            # This `bufs` array gives our custom encoder the opportunity
            # to stash additional buffers behind it. This allows us to structure
            # the data in such a way that msgpack is encoding directly from our
            # original backing buffers, instead of copying them into a new type
            # that msgpack understands. This is a significant performance gain.
            return bufs
```